### ISSUE TITLE
specialArgs usage

### ISSUE THREAD

**Issue by Reckfullies** [snowfallorg/lib#46]
I'm trying to use specialArgs on systems and homes, but I'm unsure if it is working properly.

For systems, I can just do this:
```nix
systems.hosts.my-host.specialArgs = {
    my-custom-value = "my-value";
};
```
Then access it inside "systems/x86_64-linux/my-host/default.nix" via the function parameters like this:
```nix
{ config ,lib, pkgs, my-custom-value, ... }:

{
  ...
}
```
And that works fine.

However, if I try to do the same thing for homes, it does not work at all. I can do this as stated in the docs:
```nix
homes.users."my-user@my-host".specialArgs = {
    my-custom-value = "my-value";
};
```

But if I try to access it inside "homes/x86_64-linux/my-user@my-host/default.nix" via a function parameter like I did with the system, it just gives me `error: attribute 'my-custom-value' missing` presumably because specialArgs isn't being appended to the set.

I'm not super good at understanding nix, but glancing at the implementation it does seem that specialArgs is being appended to home-manager's extraSpecialArgs inside create-home. So I don't see why this wouldn't work?

Is this a bug or am I just doing something wrong/misunderstanding something?

**Comment by jakehamilton** [comment]
Hmm, I have a hunch this has to do with Home Manager's weird handling of module args. If you can, try checking to see whether this argument is available in home modules rather than your user config.

I'll try and check on this the next time I get a chance.

**Comment by Reckfullies** [comment]
> If you can, try checking to see whether this argument is available in home modules rather than your user config.

That doesn't seem to work either. Trying to access it in a home module gives `error: function 'anonymous lambda' called without required argument 'my-custom-value'`

**Comment by Reckfullies** [comment]
I think I've figured out the issue.

Currently, snowfall-lib automatically adds the home-manager NixOS module to all systems if it is present in the flake's inputs.

I've been using standalone home-manager because I want my home-manager config to be separate from my systems, and that is what causes problems. You can't use standalone and the NixOS module at the same time, but snowfall-lib silently adds it no matter what. You can't remove home-manager from inputs either as snowfall-lib obviously needs it to be present when using standalone homes. So they will always conflict.

As a result, this was preventing `programs.home-manager` from being enabled because it checks for the NixOS module and won't enable if it is detected. This was also causing home-manager's module args to break and not pass in any of the specialArgs I defined.

I removed the part that adds the home-manager NixOS module here:
https://github.com/snowfallorg/lib/blob/92803a029b5314d4436a8d9311d8707b71d9f0b6/snowfall-lib/system/default.nix#L240-L251

And now everything seems to work fine for me. All my specialArgs are passed in as expected.

I think it would be a good idea to change how this is implemented as it is pretty confusing and silently does this with no real indication it is happening, or at least have a documented option to disable it.

**Comment by jakehamilton** [comment]
I don't think this is right, it should be possible to disable the user home via the NixOS module which will then remove any usage of home-manager on the system side.

Can you share your flake in a git repo so I can take a look?

**Comment by Reckfullies** [comment]
Alright, sorry about that. It seems that I was very mistaken, and checking just now the changes I made cause everything to break and the system doesn't even build. Not sure how it worked when I tested it yesterday, I must have messed up and ran the wrong thing. This has been confusing me a bit.

My flake is incomplete and still has a lot of private stuff not in secrets, I'd rather not share it. I just created a new basic flake directly following the docs and confirmed the issue still happens. https://github.com/Reckfullies/testflake

When using the standalone `home-manager switch --flake .#my-user@my-system` it builds fine. But if I try to do `sudo nixos-rebuild switch --flake .#my-system`, I get the error about the home's default.nix not having the custom arg I created.

I wanted to use `programs.home-manager.enable` inside the home's default.nix so that home-manager can manage itself, and I could use `home-manager switch` whenever I update my home config instead of needing to use `nixos-rebuild switch`. But it seems that doesn't work either, the home-manager executable is not added to PATH and the home won't reactivate after a reboot. The home is still managed by the NixOS module (I think?) and only properly updates and activates after rebooting when I do it via nixos-rebuild.

I apologize if what I'm saying/doing makes no sense or is wrong, I'm still quite new to learning Nix and have no clue about how some of this stuff works lol

**Comment by jakehamilton** [comment]
Thanks for the reproduction @Reckfullies! I was able to see the error you were running into and have pushed a new commit to the `dev` branch of Snowfall Lib. Can you try using the input `github:snowfallorg/lib/dev` and rebuild your system? That should have everything working for you. (I also noticed that you're setting `nixpkgs.config` in your system configuration and will need to remove that line due to how Snowfall Lib works)

**Comment by jakehamilton** [comment]
Going to close this out for now. If it is still an issue on v3 then feel free to reopen this issue :+1:
